crates/radix-engine-toolkit/src/models/canonical_address_types.rs

```diff
@@ -0,0 +1,342 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::utils::{
+    network_definition_from_network_id, network_id_from_address_string,
+};
+use paste::paste;
+use radix_engine_common::prelude::*;
+use scrypto::prelude::AddressBech32Decoder;
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+
+#[derive(Debug)]
+pub enum CanonicalAddressError {
+    FailedToDecodeBech32,
+    FailedToEncodeBech32,
+}
+
+// Required for serde Serializer
+impl Display for CanonicalAddressError {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "{:?}", self)
+    }
+}
+
+pub type NetworkId = u8;
+
+pub trait CanonicalAddress: FromStr + std::fmt::Display {
+    fn entity_type(&self) -> EntityType;
+    fn network_id(&self) -> NetworkId;
+    fn to_bech32(&self) -> Result<String, CanonicalAddressError>;
+}
+pub trait CanonicalAddressResourceType {
+    fn is_fungible(&self) -> bool;
+}
+
+// Macro which declares dedicated Canonical Addres types.
+// Arguments:
+//  - $name: used for composition of the new type name Canonical_NAME_Address
+//  - $entity_type: pattern of valid entity types for this new type
+//  - $fungibility_validator: optional, if specified additionally CanonicalAddressResourceType trait
+//    is declared, this expects closure with parameter &NodeId which can be used to validate type of resource
+macro_rules! make_canonical_address {
+
+    ($name: ident, $entity_type: pat) => {
+        paste! {
+            #[derive(
+                Clone, Debug, PartialEq, Eq, Hash, SerializeDisplay, DeserializeFromStr,
+            )]
+            pub struct [<Canonical $name Address>] {
+                address: NodeId,
+                network_id: NetworkId,
+            }
+
+            impl [<Canonical $name Address>] {
+                pub fn try_from_global_address(
+                    global_address: &GlobalAddress,
+                    network_id: NetworkId,
+                ) -> Option<Self> {
+                    Self::try_from_node_id(&global_address.into_node_id(), network_id)
+                }
+
+                pub fn try_from_internal_address(
+                    internal_address: &InternalAddress,
+                    network_id: NetworkId,
+                ) -> Option<Self> {
+                    Self::try_from_node_id(&internal_address.into_node_id(), network_id)
+                }
+
```

Comment:
I think that a better way to do this would be to implement `TryFrom<T>` where T is any type that implemens `Into<NodeId>`.

Reply:
We can't do it this way since we need the network id. I've refactored this to be the following:

```rust
pub fn new(node_id: impl Into<NodeId>, network_id: NetworkId) -> Result<Self, CanonicalAddressError>
```

Which I think is a nice generic interface for constructing those types.